### PR TITLE
Add umpire pools to face geometric factor memory

### PR DESCRIPTION
**Pull request by artv3**
I'm not 100% sure if I did this right, @tomstitt , @v-dobrev could I trouble you for a review? 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2934](https://github.com/mfem/mfem/pull/2934) | @artv3 | @v-dobrev | @v-dobrev + @camierjs | 3/28/22 | 6/16/22 | 6/23/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tomstitt** [comment]
I think it would be better to make the device type an optional argument like we did with the geometric factors: https://github.com/mfem/mfem/blob/master/mesh/mesh.cpp#L11935-L11978.

You can use `MemoryType::DEFAULT` like we do [here](https://github.com/mfem/mfem/blob/master/mesh/mesh.hpp#L1818-L1838) then pick the actual type based on that value like we do [here](https://github.com/mfem/mfem/blob/master/mesh/mesh.cpp#L11973-L11974). Does that make sense?

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -12021,6 +12021,9 @@ FaceGeometricFactors::FaceGeometricFactors(const Mesh *mesh,
    IntRule = &ir;
    computed_factors = flags;
 
+   MemoryType d_mt = Device::Allows(Backend::RAJA_CUDA) == true ?
```
Comment (camierjs) [inline mesh/mesh.cpp:12024]: We should use an agnostic way of using the non-default memory type, as suggested/done in the `geometricFactors`.

Comment (pazner) [reply mesh/mesh.cpp:12024]: @artv3, could you take a look at this suggestion ☝️

Comment (artv3) [reply mesh/mesh.cpp:12024]: Yes! Sorry for the delay, it is on my list.

Comment (artv3) [reply mesh/mesh.cpp]: Thanks, just refactored it like in geometricFactors.

File: mesh/mesh.cpp [hunk 2]
```diff
@@ -12050,28 +12052,32 @@ FaceGeometricFactors::FaceGeometricFactors(const Mesh *mesh,
                                           ElementDofOrdering::LEXICOGRAPHIC,
                                           type,
                                           L2FaceValues::SingleValued );
-   Vector Fnodes(face_restr->Height());
+   Vector Fnodes(face_restr->Height(), d_mt);
    face_restr->Mult(*nodes, Fnodes);
 
    unsigned eval_flags = 0;
+
+   MemoryType my_d_mt = (d_mt != MemoryType::DEFAULT) ? d_mt :
```
Comment (camierjs) [inline mesh/mesh.cpp:12060]: I have a preference for less generic names than `my_***`, but I see that this file is already using such prefix, so I'm fine with it.

**Review by camierjs — APPROVED** [review]

File: mesh/mesh.cpp [hunk 3]
```diff
@@ -12050,28 +12052,32 @@ FaceGeometricFactors::FaceGeometricFactors(const Mesh *mesh,
                                           ElementDofOrdering::LEXICOGRAPHIC,
                                           type,
                                           L2FaceValues::SingleValued );
-   Vector Fnodes(face_restr->Height());
+   Vector Fnodes(face_restr->Height(), d_mt);
```
Comment (artv3) [inline mesh/mesh.cpp:12055]:
```suggestion
   Vector Fnodes(face_restr->Height(), my_d_mt);
```

**Comment by artv3** [comment]
Hey... this failed: 
 84/112 Test  #84: navier_tgv_np2 ........................***Failed    0.37 sec

**Comment by v-dobrev** [comment]
> Hey... this failed: 84/112 Test ` #84`: navier_tgv_np2 ........................***Failed 0.37 sec

I'll try restarting the failed windows job.

**Comment by tzanio** [comment]
@artv3, try re-merging `master`, there was a cmake issues that caused such failures...

**Comment by v-dobrev** [comment]
Now a different test failed in that windows job -- it looks like that windows CI environment is not very stable.

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**